Add rudimentary support for LMS signing

mateusz-mowka opened the pull request:
To use LMS:

1. Clone hash-sigs: https://github.com/cisco/hash-sigs and build the "demo" tool
2. Make sure "demo" is in your path, for example copy it to /usr/bin
3. Generate unsigned LCP list, for example like this:
    ./lcp2_mlehash --create --cmdline "logging=serial,memory,vga" --alg sha256 ../tboot/tboot.gz > tboot_hash.hash
    ./lcp2_crtpolelt --create --type mle2 --minver 0x00 --alg sha256 --ctrl 0x00 tboot_hash.hash --out tboot_mle2.elt
    ./lcp2_crtpollist --create --listver 0x300 --out tboot_lcp_list.lst tboot_mle2.elt

4. Create key pair:
    demo genkey <keyname> 20/4
5. Copy list file:
    cp tboot_lcp_list.lst lms_signed_tboot_lcp_list.lst
6. Sign:
    ./lcp2_crtpollist --verbose --sign --sigalg lms --hashalg sha256 --pub <keyname.pub> --priv <keyname.priv> --out lms_signed_tboot_lcp_list.lst

During execution new files will be created: lcp_list_digest, which contains LCP list data digest, which will be signed, and lcp_list_digest.sig which contains the signature.

The way it works is that lcp2_crtpollist will call:

demo sign <keyname> lcp_list_digest

And then read the signature that "demo" generated and build lcp_signature_2_1 structure based on that.

lcp2_crtpollist --show also works with LMS.

mateusz-mowka committed:
Add rudimentary support for LMS signing

mateusz-mowka commented on lcptools-v2/hash.c:
```diff
@@ -85,62 +78,33 @@ bool hash_buffer(const unsigned char* buf, size_t size, tb_hash_t *hash,
+    
+    EVP_DigestInit(ctx, md);
+    EVP_DigestUpdate(ctx, buf, size);
+    EVP_DigestFinal(ctx, hash->sha1, NULL);
```
This is union so it doesn't matter which member receives the digest

mateusz-mowka commented:
TODO:

1. Hybrid signature support
2. Support for TPM_ALG_LMS when calculating allowed signature algs
3. Hashing LMS public key to generate signed LCP policy hash

mateusz-mowka committed:
Update lcp3.h with LMS definitions

apawlickix commented on include/lcp3.h:
```diff
@@ -408,9 +411,96 @@ typedef struct __packed {
+    uint8_t  Version;
+    uint16_t KeySize;
+    uint16_t HashAlg;
+    lms_signature_block signature;
```
Consistency issue, signature->Signature

mateusz-mowka replied:
Done

apawlickix commented on include/lcp3.h:
```diff
@@ -408,9 +411,96 @@ typedef struct __packed {
+} rsa_lms_key_and_signature;
+
+typedef struct __packed {
+    uint8_t  Version; //bit7 of this
```
Half of the comment gone

apawlickix commented on lcptools-v2/hash.c:
```diff
@@ -85,62 +78,33 @@ bool hash_buffer(const unsigned char* buf, size_t size, tb_hash_t *hash,
+    EVP_MD_CTX *ctx = EVP_MD_CTX_create();
+    const EVP_MD *md;
 
+    if ( hash_alg == TB_HALG_SHA1 || hash_alg == TB_HALG_SHA1_LG ) {
```
If we're anyway touching this part, couldn't we replace it with switch?

apawlickix replied:
@mateusz-mowka

apawlickix commented on lcptools-v2/lcputils.c:
```diff
@@ -190,6 +191,7 @@ bool write_file(const char *file, const void *data, size_t size)
                 file, strerror(errno));
         return false;
     }
+    fseek(fp, offset, SEEK_SET);
```
Check if successful?

apawlickix commented on lcptools-v2/pollist2_1.c:
```diff
@@ -270,6 +314,21 @@ Out: Pointer to an empty structure
+                                                                KeyAndSignature);
+
+    lcp_signature_2_1 *sig = calloc(sig_size, 1);
+    if (sig == NULL) {
+        return NULL;
+    }
+    return sig;
```
Isn't it redundant? We'd return null anyway, without this check.

apawlickix replied:
@mateusz-mowka

mateusz-mowka committed:
Enhance LMS support and improve display formatting in policy elements

- Added LMS signature masks and definitions in lcp3.h and lcputils.h.
- Added LMS public key hashing to generate policy hash
- Updated logging for LCP version in crtpol.c.
- Improved display formatting for hash outputs across multiple files.

mateusz-mowka commented:
Left to do:

1. Signature verification

mateusz-mowka committed:
Refactor LMS signature structure and add LMS verification (WIP)

mateusz-mowka committed:
Add hashing to the verification of policy list LMS signature

marziel-intel committed:
Change tb_hash_t malloc in pconf_legacy.c

marziel-intel committed:
Change tb_hash_t malloc in lcputils.c

marziel-intel committed:
Change tb_hash_t malloc in pconf2_elt.c

marziel-intel committed:
Fix wrong offset in pollist2_1.c

mateusz-mowka committed:
Remove reading mode from write only files.
Close file handle before performing signing to fix
signature verification issues.

mateusz-mowka commented on lcptools-v2/hash.c:
```diff
@@ -85,62 +78,33 @@ bool hash_buffer(const unsigned char* buf, size_t size, tb_hash_t *hash,
+    
+    EVP_DigestInit(ctx, md);
+    EVP_DigestUpdate(ctx, buf, size);
+    EVP_DigestFinal(ctx, hash->sha1, NULL);
```
This is a union, so all members point to the same address. The suggestion is unnecessary bloat.

apawlickix reviewed (approved):
Some nitpicks added
